core/test/Cardano/Address/ScriptSpec.hs

```diff
@@ -251,6 +261,34 @@ spec = do
                         ]
                     ]
             validateScript script `shouldBe` Right ()
+
+    describe "can perform roundtrip JSON serialization & deserialization" $
+        it "fromJSON . toJSON === pure" $ do
+            let roundtrip =
+                    Json.decode . Json.encode . toJSON :: Script -> Maybe Script
+            property $ \(t :: Script) -> roundtrip t === pure t
+
   where
     toHexText = T.decodeUtf8 . encode EBase16
     toHexText' (ScriptHash bytes) = toHexText bytes
+
+instance Arbitrary Script where
+    arbitrary = do
+        reqAllGen <- do
+            n <- choose (1,10)
+            pure $ RequireAllOf <$> vector n
+        reqAnyGen <- do
+            n <- choose (1,10)
+            pure $ RequireAnyOf <$> vector n
+        reqMofNGen <- do
+            m <- choose (2,5)
+            n <- choose ((fromInteger $ toInteger m),10)
+            pure $ RequireMOf m <$> vector n
+        let reqSig =
+                (RequireSignatureOf . KeyHash . BS.pack) <$> replicateM 28 arbitrary
+        oneof
+            (replicate 15 reqSig ++
+            [ reqAllGen
+            , reqAnyGen
+            , reqMofNGen
+            ])
```

Comment:
I looked it up... We do need to be careful to avoid generators which fail to terminate, or produce very large results. [Generating Recursive Data Types](http://www.cse.chalmers.se/~rjmh/QuickCheck/manual_body.html#16) :thinking:.

Also, can `shrink = genericShrink` work here?

```suggestion

instance Arbitrary Script where
    arbitrary = sized script'
      where
        script' 0 = RequireSignatureOf . KeyHash . BS.pack <$> replicateM 28 arbitrary
        script' n | n>0 = oneof
            [ RequireAllOf <$> listOf arbitrary
            , RequireAnyOf <$> listOf arbitrary
            , RequireMOf <$> choose (2,5) <*> listOf arbitrary
                -- note: perhaps the last one is too invalid
            ]
```

Reply:
```
instance Arbitrary Script where
    arbitrary = sized script'
      where
        script' 0 = RequireSignatureOf . KeyHash . BS.pack <$> replicateM 28 arbitrary
        script' n =
            let n' = n `div` 2
                scripts = listOf (script' n')
                conds = [ RequireAllOf <$> scripts
                        , RequireAnyOf <$> scripts ]
            in if n' >= 5 then
                   oneof (conds ++ [RequireSomeOf <$> choose (2,5) <*> scripts])
               else
                   oneof conds

```
seems to stuck when running... on the other hand, for me, it is much less clear to read

Reply:
OK I fixed it. I think it's important to learn how to generate tests properly and use best practices for quickcheck.